<h1>Оглавление.<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Анализ-базы-данных-сервиса-для-чтения-книг." data-toc-modified-id="Анализ-базы-данных-сервиса-для-чтения-книг.-1">Анализ базы данных сервиса для чтения книг.</a></span><ul class="toc-item"><li><span><a href="#1.-Описание-данных." data-toc-modified-id="1.-Описание-данных.-1.1">1. Описание данных.</a></span></li><li><span><a href="#2.-Ознакомление-с-данными." data-toc-modified-id="2.-Ознакомление-с-данными.-1.2">2. Ознакомление с данными.</a></span></li></ul></li><li><span><a href="#3.-Выполнения-заданий." data-toc-modified-id="3.-Выполнения-заданий.-2">3. Выполнения заданий.</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#3.1.-Посчитайте,-сколько-книг-вышло-после-1-января-2000-года." data-toc-modified-id="3.1.-Посчитайте,-сколько-книг-вышло-после-1-января-2000-года.-2.0.1">3.1. Посчитайте, сколько книг вышло после 1 января 2000 года.</a></span></li><li><span><a href="#3.2.-Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку." data-toc-modified-id="3.2.-Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку.-2.0.2">3.2. Для каждой книги посчитайте количество обзоров и среднюю оценку.</a></span></li><li><span><a href="#3.3.-Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры." data-toc-modified-id="3.3.-Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры.-2.0.3">3.3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.</a></span></li><li><span><a href="#3.4.-Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками." data-toc-modified-id="3.4.-Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками.-2.0.4">3.4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.</a></span></li><li><span><a href="#3.5.-Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок." data-toc-modified-id="3.5.-Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок.-2.0.5">3.5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a></span></li></ul></li><li><span><a href="#Вывод." data-toc-modified-id="Вывод.-2.1">Вывод.</a></span><ul class="toc-item"><li><span><a href="#Чек-лист-доработок-" data-toc-modified-id="Чек-лист-доработок--2.1.1">Чек-лист доработок </a></span></li></ul></li></ul></li></ul></div>

# Анализ базы данных сервиса для чтения книг.

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания**
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

## 1. Описание данных.

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

**Как выполнить задание?**

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

## 2. Ознакомление с данными.

In [1]:
# импортируем библиотеки
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

In [3]:
con = create_engine('postgresql+psycopg2://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db']))

In [4]:
# обернём в функцию для удобства чтения
def select(sql):
    return pd.read_sql(sql,con)

# 3. Выполнения заданий.

### 3.1. Посчитайте, сколько книг вышло после 1 января 2000 года.

In [5]:
books = '''
SELECT 
count(books.book_id) AS books_cnt

FROM 
books

WHERE 
books.publication_date > '2000-01-01'::date
'''

In [6]:
select(books)

books_cnt
0        819

После 1 января 2000 года было выпущено 819 книг.

### 3.2. Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [7]:
reviews = '''
SELECT 
    books.title AS title,
    COUNT(DISTINCT reviews.text) AS cnt_reviews,
    AVG(ratings.rating) AS avg_rating
    
FROM
    books
LEFT JOIN ratings ON ratings.book_id = books.book_id
LEFT JOIN reviews ON reviews.book_id = books.book_id

GROUP BY
     books.title
     
--HAVING
--avg(ratings.rating) = 5
--COUNT(DISTINCT reviews.text) = 0

order by
     cnt_reviews DESC,
     avg_rating DESC
     
limit 10
'''

In [8]:
select(reviews)

title  cnt_reviews  avg_rating
0                                Memoirs of a Geisha            8    4.138462
1                            Twilight (Twilight  #1)            7    3.662500
2  Harry Potter and the Prisoner of Azkaban (Harr...            6    4.414634
3  Harry Potter and the Chamber of Secrets (Harry...            6    4.287500
4                                     The Book Thief            6    4.264151
5                                   The Glass Castle            6    4.206897
6                The Hobbit  or There and Back Again            6    4.125000
7                          Outlander (Outlander  #1)            6    4.125000
8  The Curious Incident of the Dog in the Night-Time            6    4.081081
9  The Lightning Thief (Percy Jackson and the Oly...            6    4.080645

- Для 6 книг отсутсвуют обзоры.
- Больше всего обзоров у книги **Memoirs of a Geisha**.
- У 44 книг рейтинг - 5.0, при этом количество обзоров от 0 до 4. С 4 обзорами и рейтингом 5.0 книга **A Dirty Job (Grim Reaper #1)**.

### 3.3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [9]:
publishers = '''
SELECT 
count(books.book_id),
publishers.publisher

FROM 
books

LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id

WHERE 
books.num_pages > 50

GROUP BY 
publishers.publisher

ORDER BY count DESC

LIMIT 10
'''

In [10]:
select(publishers)

count                  publisher
0     42              Penguin Books
1     31                    Vintage
2     25   Grand Central Publishing
3     24           Penguin Classics
4     19           Ballantine Books
5     19                     Bantam
6     17                    Berkley
7     14         St. Martin's Press
8     14              Berkley Books
9     13  William Morrow Paperbacks

Издательство *Penguin Books* - выпустило наибольшее число книг толще 50 страниц (42 шт).

 ### 3.4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [11]:
test = '''SELECT
      tab.author AS author,
      AVG(tab.avg_rating) AS avg_rating
FROM
(SELECT
     authors.author AS author,
     books.book_id AS id,
     AVG(ratings.rating) AS avg_rating,
     COUNT(ratings.rating) AS cnt_rating
FROM
    authors
LEFT JOIN books ON books.author_id = authors.author_id
LEFT JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    author,
    id
    HAVING
    COUNT(ratings.rating) >= 50) AS tab
GROUP BY
    author
ORDER BY
    avg_rating DESC
'''

In [12]:
select(test)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

Самая высокая средняя оценка у автора *J.K. Rowling/Mary GrandPré* - 4.28.

### 3.5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
review_avg = '''
SELECT 
avg(reviews.review_cnt) AS review_avg

FROM (SELECT reviews.username,
      count(reviews.text) AS review_cnt
      FROM reviews
      GROUP BY reviews.username) AS reviews

WHERE 
reviews.username in (SELECT ratings.username
                    FROM ratings
                    GROUP BY ratings.username
                    HAVING count(ratings.rating) > 50);
'''

In [14]:
select(review_avg)

review_avg
0   24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3.

## Вывод.

Результаты анализа базы данных сервиса для чтения книг по подписке:
- После 1 января 2000 года было выпущено 819 книг.
- Для 6 книг отсутсвуют обзоры.
- Больше всего обзоров у книги **Memoirs of a Geisha**.
- У 44 книг рейтинг - 5.0, при этом количество обзоров от 0 до 4. С 4 обзорами и рейтингом 5.0 книга **A Dirty Job (Grim Reaper #1)**.
- Издательство Penguin Books - выпустило наибольшее число книг толще 50 страниц (42 шт).
- Самая высокая средняя оценка у автора *J.K. Rowling/Mary GrandPré* - 4.28.
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.3.